### PR TITLE
Add optional executor restriction to cw3-flex

### PR DESCRIPTION
**Pull request by ueco-jb**
closes https://github.com/CosmWasm/cw-plus/issues/739

Test are bloated, but all of them are... Some proper refactoring would be useful here.

### CODE REVIEW

File: contracts/cw3-flex-multisig/src/contract.rs [hunk 1]
```diff
@@ -192,15 +193,33 @@ pub fn execute_execute(
...
         return Err(ContractError::WrongExecuteStatus {});
     }
 
+    // Executor can be set in 3 ways:
+    // - Member: any member of the voting group can execute
+    // - Only: only passed address is able to execute
+    // - None: Anyone can execute message
+    let cfg = CONFIG.load(deps.storage)?;
+    if let Some(executor) = cfg.executor {
+        match executor {
```
Comment (ethanfrey) [inline contracts/cw3-flex-multisig/src/contract.rs:209]: Is this something that could be pulled out to a method on Executor?
Already thinking of reusing it for who can make proposals. (but default is Member)

File: contracts/cw3-flex-multisig/src/contract.rs [hunk 2]
```diff
@@ -1191,6 +1226,128 @@ mod tests {
...
+            Addr::unchecked(Addr::unchecked(VOTER2)),
+            flex_addr,
+            &execution,
+            &[],
+        )
+        .unwrap();
+    }
+
+    #[test]
+    fn execute_with_executor_only() {
```
Comment (ethanfrey) [inline contracts/cw3-flex-multisig/src/contract.rs:1286]: Nice tests here. And yes, some of that Suite goodness from poe-contracts would be great here

File: contracts/cw3-flex-multisig/src/state.rs [hunk 3]
```diff
@@ -1,16 +1,29 @@
...
 use serde::{Deserialize, Serialize};
 
+use cosmwasm_std::Addr;
 use cw4::Cw4Contract;
 use cw_storage_plus::Item;
 use cw_utils::{Duration, Threshold};
 
+/// Defines who is able to execute proposals once passed
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub enum Executor {
```
Comment (ethanfrey) [inline contracts/cw3-flex-multisig/src/state.rs:11]: I went to the issue to add a comment, then saw you had implemented it (thank you!)

My idea was to only use `Option<Executor>` in the InstantiateMsg and store `Executor` in Config.
This also means to add one more enum variant: `Everyone`.

~~When instantiating, if the message has `None`, we assign the default `Executor::Everyone`, but then the application of the default / meaning of "None" is made on init and not later~~ (see below)

File: contracts/cw3-flex-multisig/src/state.rs [hunk 4]
```diff
@@ -1,16 +1,29 @@
...
+
 #[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
 pub struct Config {
     pub threshold: Threshold,
     pub max_voting_period: Duration,
     // Total weight and voters are queried from this contract
     pub group_addr: Cw4Contract,
+    // who is able to execute passed proposals
+    // None means that anyone can execute
+    pub executor: Option<Executor>,
```
Comment (ethanfrey) [inline contracts/cw3-flex-multisig/src/state.rs:26]: Disregard the above comment. This is great as it is drop-in state compatible with previous version, not requiring an explicit migration (we should have some placeholder migrate function, but no state change needed). No need to make breaking changes for some aesthetic opinion of mine.

**Review by ethanfrey — APPROVED** [review]
Wow. Thanks for the quick reaction here.

Code looks good. I had an idea how to extend it a bit, left comments on the Executor variant and making the check a method.

The other idea (for a different PR) was to use the same pattern for who can submit a proposal, but have that default to "Member" rather than "Everyone" if the InstantiateMsg has `None` in the field.

File: contracts/cw3-flex-multisig/src/contract.rs (hunk 1 above)
Comment (ueco-jb) [reply contracts/cw3-flex-multisig/src/contract.rs]: I moved it under `Config` struct. I think it's more versatile this way.
What do you think?

File: contracts/cw3-flex-multisig/src/state.rs (hunk 4 above)
Comment (ueco-jb) [reply contracts/cw3-flex-multisig/src/state.rs:28]: Yes, I considered that but as you pointed out - this way it's completely backward compatible.

File: contracts/cw3-flex-multisig/src/contract.rs [hunk 5]
```diff
@@ -200,25 +200,8 @@ pub fn execute_execute(
...
-                    .ok_or(ContractError::Unauthorized {})?;
-            }
-            Executor::Only(addr) => {
-                if addr != info.sender {
-                    return Err(ContractError::Unauthorized {});
-                }
-            }
-        }
-    }
+    cfg.authorize(&deps.querier, &info.sender)?;
```
Comment (ethanfrey) [inline contracts/cw3-flex-multisig/src/contract.rs:204]: nice

File: contracts/cw3-flex-multisig/src/state.rs [hunk 6]
```diff
@@ -26,5 +28,29 @@ pub struct Config {
     pub executor: Option<Executor>,
 }
 
+impl Config {
```
Comment (ethanfrey) [inline contracts/cw3-flex-multisig/src/state.rs:31]: looks good

**Review by ethanfrey — APPROVED** [review]
thanks

**Pull request merged**